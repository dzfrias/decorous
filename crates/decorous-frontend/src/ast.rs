use std::{borrow::Cow, fmt, path::Path};

use heck::ToSnekCase;
use itertools::Itertools;
use rslint_parser::SyntaxNode;

use crate::{css::ast::Css, location::Location};

/// The abstract syntax tree representation of decorous source code.
///
/// This struct has three main components (all with respective getter methods):
/// 1. [`nodes`](Self::nodes), for holding HTML markup.
/// 2. [`script`](Self::script), for holding JavaScript.
/// 3. [`css`](Self::css), for holding CSS.
///
/// One of the best uses of `DecorousAst` is to hold data, read some parts, and eventually use
/// [`into_components()`](DecorousAst::into_components()) when the time has come to apply more
/// complex transformations that require ownership.
#[derive(Debug)]
pub struct DecorousAst<'a> {
    pub nodes: Vec<Node<'a, Location>>,
    pub script: Option<SyntaxNode>,
    pub css: Option<Css>,
    pub wasm: Option<Code<'a>>,
    pub comptime: Option<Code<'a>>,
}

/// A node of the [AST](DecorousAst).
///
/// It contains [metadata](`Self::metadata()`) (of type `T`), and the
/// actual node data, retrieved by [`node_type()`](`Self::node_type()`).
#[derive(Debug, Clone, PartialEq)]
pub struct Node<'a, T> {
    pub metadata: T,
    pub node_type: NodeType<'a, T>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum NodeType<'a, T> {
    Element(Element<'a, T>),
    Text(Text<'a>),
    Comment(Comment<'a>),
    SpecialBlock(SpecialBlock<'a, T>),
    Mustache(Mustache),
}

#[derive(Debug, Clone, PartialEq, Hash)]
pub struct Mustache(pub SyntaxNode);

impl fmt::Display for Mustache {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::ops::Deref for Mustache {
    type Target = SyntaxNode;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Copy)]
pub struct Text<'a>(pub &'a str);

impl fmt::Display for Text<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl<'a> std::ops::Deref for Text<'a> {
    type Target = &'a str;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Copy)]
pub struct Comment<'a>(pub &'a str);

impl<'a> std::ops::Deref for Comment<'a> {
    type Target = &'a str;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Element<'a, T> {
    pub tag: &'a str,
    pub attrs: Vec<Attribute<'a>>,
    pub children: Vec<Node<'a, T>>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum SpecialBlock<'a, T> {
    For(ForBlock<'a, T>),
    If(IfBlock<'a, T>),
    Use(UseBlock<'a>),
}

#[derive(Debug, Clone, PartialEq)]
pub struct ForBlock<'a, T> {
    pub binding: &'a str,
    pub index: Option<&'a str>,
    pub expr: SyntaxNode,
    pub inner: Vec<Node<'a, T>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct IfBlock<'a, T> {
    pub expr: SyntaxNode,
    pub inner: Vec<Node<'a, T>>,
    pub else_block: Option<Vec<Node<'a, T>>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct UseBlock<'a> {
    pub path: &'a Path,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Attribute<'a> {
    EventHandler(EventHandler<'a>),
    Binding(&'a str),
    KeyValue(&'a str, Option<AttributeValue<'a>>),
}

#[derive(Debug, Clone, PartialEq)]
pub struct EventHandler<'a> {
    pub event: &'a str,
    pub expr: SyntaxNode,
}

#[derive(Debug, Clone, PartialEq)]
pub enum AttributeValue<'a> {
    Literal(Cow<'a, str>),
    JavaScript(SyntaxNode),
}

#[derive(Debug, Clone, PartialEq)]
pub enum CollapsedChildrenType<'a> {
    Text(&'a str),
    Html(String),
}

#[derive(Debug)]
pub struct Code<'a> {
    pub lang: &'a str,
    pub body: &'a str,
    pub offset: usize,
    pub comptime: bool,
}
impl<'a, T> Element<'a, T> {
    pub fn descendents(&'a self) -> NodeIter<'a, T> {
        NodeIter::new(&self.children)
    }

    pub fn descendents_mut<F>(&'a mut self, f: &mut F)
    where
        F: FnMut(&mut Node<'a, T>),
    {
        for child in &mut self.children {
            f(child);
            if let NodeType::Element(elem) = &mut child.node_type {
                elem.descendents_mut(f);
            }
        }
    }

    pub fn has_immediate_mustache(&self) -> bool {
        self.children
            .iter()
            .any(|child| matches!(child.node_type, NodeType::Mustache(_)))
    }

    pub fn js_valid_tag_name(&self) -> Cow<'a, str> {
        if self.tag.contains('-') {
            Cow::Owned(self.tag.to_snek_case())
        } else {
            Cow::Borrowed(self.tag)
        }
    }
}

impl<'a, T> Node<'a, T> {
    pub fn new(ty: NodeType<'a, T>, metadata: T) -> Self {
        Self {
            metadata,
            node_type: ty,
        }
    }

    /// Recursively cast each the metadata field of each node into a new type. The provided
    /// function receives the previous metadata of the node, and should return the corresponding new
    /// metadata.
    pub fn cast_meta<U, F>(self, transfer_func: &mut F) -> Node<'a, U>
    where
        F: FnMut(&Node<'a, T>) -> U,
    {
        macro_rules! cast_children {
            ($children_vec:expr, $transfer:expr) => {
                $children_vec
                    .into_iter()
                    .map(|node| node.cast_meta($transfer))
                    .collect()
            };
        }

        let new_meta = transfer_func(&self);

        // A bit verbose, but it's the only way to do a full type cast throughout the entire AST
        match self.node_type {
            NodeType::SpecialBlock(block) => Node {
                metadata: new_meta,
                node_type: NodeType::SpecialBlock(match block {
                    SpecialBlock::If(if_block) => SpecialBlock::If(IfBlock {
                        expr: if_block.expr,
                        inner: cast_children!(if_block.inner, transfer_func),
                        else_block: if_block
                            .else_block
                            .map(|nodes| cast_children!(nodes, transfer_func)),
                    }),
                    SpecialBlock::For(for_block) => SpecialBlock::For(ForBlock {
                        inner: cast_children!(for_block.inner, transfer_func),
                        binding: for_block.binding,
                        index: for_block.index,
                        expr: for_block.expr,
                    }),
                    SpecialBlock::Use(use_block) => SpecialBlock::Use(use_block),
                }),
            },
            NodeType::Element(elem) => Node {
                metadata: new_meta,
                node_type: NodeType::Element(Element {
                    tag: elem.tag,
                    attrs: elem.attrs,
                    children: cast_children!(elem.children, transfer_func),
                }),
            },
            NodeType::Text(text) => Node {
                metadata: new_meta,
                node_type: NodeType::Text(text),
            },
            NodeType::Comment(comment) => Node {
                metadata: new_meta,
                node_type: NodeType::Comment(comment),
            },
            NodeType::Mustache(syntax_node) => Node {
                metadata: new_meta,
                node_type: NodeType::Mustache(syntax_node),
            },
        }
    }
}

impl<'a, T> IfBlock<'a, T> {
    pub fn inner_recursive(&'a self) -> NodeIter<'a, T> {
        NodeIter::new(&self.inner)
    }

    pub fn else_recursive(&'a self) -> Option<NodeIter<'a, T>> {
        self.else_block
            .as_ref()
            .map(|else_block| NodeIter::new(else_block))
    }
}

impl<'a> DecorousAst<'a> {
    /// Creates a recursive iterator over the nodes of the template.
    pub fn iter_nodes(&'a self) -> NodeIter<'a, Location> {
        NodeIter::new(&self.nodes)
    }
}

pub fn traverse_with<'a, T, F, G>(nodes: &'a [Node<'a, T>], predicate: &mut F, body_func: &mut G)
where
    F: FnMut(&Element<'a, T>) -> bool,
    G: FnMut(&Node<'a, T>),
{
    for node in nodes {
        body_func(node);
        if let NodeType::Element(elem) = &node.node_type {
            if !predicate(elem) {
                continue;
            }
            traverse_with(&elem.children, predicate, body_func);
        }
    }
}

pub fn traverse_mut<'a, T, F>(nodes: &mut [Node<'a, T>], f: &mut F)
where
    F: FnMut(&mut Node<'a, T>),
{
    for node in nodes {
        f(node);
        if let NodeType::Element(elem) = &mut node.node_type {
            traverse_mut(&mut elem.children, f);
        }
    }
}

#[derive(Debug)]
pub struct NodeIter<'a, T> {
    stack: Vec<&'a Node<'a, T>>,
}

impl<'a, T> NodeIter<'a, T> {
    pub fn new(nodes: &'a [Node<'a, T>]) -> NodeIter<'a, T> {
        let mut stack = Vec::with_capacity(nodes.len());
        stack.extend(nodes.iter().rev());
        Self { stack }
    }
}

impl<'a, T> Iterator for NodeIter<'a, T> {
    type Item = &'a Node<'a, T>;

    fn next(&mut self) -> Option<Self::Item> {
        self.stack.pop().map(|node| {
            if let NodeType::Element(elem) = &node.node_type {
                self.stack.extend(elem.children.iter().rev());
            }

            node
        })
    }
}

impl<'a, T> fmt::Display for Node<'a, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match &self.node_type {
            NodeType::Text(t) => write!(f, "{t}"),
            NodeType::Comment(Comment(c)) => write!(f, "<!--{c}-->"),
            NodeType::Element(elem) => write!(f, "{elem}"),
            NodeType::Mustache(js) => write!(f, "{{{js}}}"),
            NodeType::SpecialBlock(block) => write!(f, "{block}"),
        }
    }
}

impl<'a, T> fmt::Display for Element<'a, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "<{}{}{}>{}</{0}>",
            self.tag,
            (!self.children.is_empty())
                .then_some(" ")
                .unwrap_or_default(),
            self.attrs.iter().join(" "),
            self.children
                .iter()
                .map(|elem| format!("  {elem}"))
                .join("")
        )
    }
}

impl<'a> fmt::Display for Attribute<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Attribute::KeyValue(key, Some(val)) => write!(f, "{key}={val}"),
            Attribute::KeyValue(key, None) => write!(f, "{key}"),
            Attribute::EventHandler(event_handler) => write!(f, "{event_handler}"),
            Attribute::Binding(binding) => write!(f, "bind:{binding}"),
        }
    }
}

impl<'a> fmt::Display for AttributeValue<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            AttributeValue::Literal(literal) => write!(f, "\"{literal}\""),
            AttributeValue::JavaScript(js) => write!(f, "{{{js}}}"),
        }
    }
}

impl<'a> fmt::Display for EventHandler<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "on:{}={{{}}}", self.event, self.expr)
    }
}

impl<'a, T> fmt::Display for SpecialBlock<'a, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SpecialBlock::If(if_block) => write!(f, "{if_block}"),
            SpecialBlock::For(for_block) => write!(f, "{for_block}"),
            SpecialBlock::Use(use_block) => write!(f, "{use_block}"),
        }
    }
}

impl<'a, T> fmt::Display for IfBlock<'a, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{{#if {}}}\n{}\n{{/if}}",
            self.expr,
            self.inner.iter().map(|elem| format!("  {elem}")).join(""),
        )
    }
}

impl<'a, T> fmt::Display for ForBlock<'a, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{{#for {} in {}}}\n{}\n{{/for}}",
            self.index.map_or_else(
                || self.binding.to_owned(),
                |idx| format!("{idx}, {}", self.binding)
            ),
            self.expr,
            self.inner.iter().map(|elem| format!("  {elem}")).join(""),
        )
    }
}

impl<'a> fmt::Display for UseBlock<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{{#use \"{}\"}}", self.path.display())
    }
}
