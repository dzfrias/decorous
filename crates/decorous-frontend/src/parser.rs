use rslint_parser::SyntaxNode;
use std::{collections::VecDeque, str::Chars};
use thiserror::Error;

use crate::ast::{
    Attribute, AttributeValue, Element, EventHandler, ForBlock, IfBlock, Location, Node, NodeType,
    SpecialBlock,
};

#[derive(Debug)]
pub struct Parser<'a> {
    // Source
    source: &'a str,
    input: Chars<'a>,
    // Not using `Peekable` because access to `as_str()` of `input` is crucial.
    peek_buffer: Option<char>,

    // A queue we maintain of the closed tags we find. This allows us to find invalid closing tag
    // errors
    close_tag_queue: VecDeque<&'a str>,

    // Current character index
    index: usize,
    // Current character index, with proper utf8 length taken into account. We need this because
    // we take a slice of `source`, which operates in bytes, not code points
    slice_index: usize,

    errors: Vec<ParseError>,
}

#[derive(Debug, Error, PartialEq)]
pub enum ParseError {
    #[error("invalid closing tag, expected {0}")]
    InvalidClosingTag(String, Location),
    #[error("invalid character, expected {0}, got {1}")]
    ExpectedCharacter(char, char, Location),
    #[error("invalid character, expected {0:?}, got {1}")]
    ExpectedCharacterAny(Vec<char>, char, Location),
    #[error("expected end of comment")]
    ExpectedEndOfComment(Location),
    #[error("invalid closing block, expected {0}, got {1}")]
    InvalidClosingBlock(String, String, Location),
    #[error("expected string {0}, got {1}")]
    ExpectedString(String, String, Location),
    #[error("unrecognized special block: \"{0}\"")]
    UnrecognizedSpecialBlock(String, Location),
}

/// Zero-copy parser that takes in decorous HTML syntax.
impl<'a> Parser<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            source: input,
            input: input.chars(),
            peek_buffer: None,

            close_tag_queue: VecDeque::new(),

            index: 0,
            slice_index: 0,

            errors: vec![],
        }
    }

    /// Parse the received input into a collection of nodes. Also, return the errors found while
    /// parsing.
    pub fn parse(mut self) -> (Vec<Node<'a>>, Vec<ParseError>) {
        let nodes = self.parse_nodes();

        (nodes, self.errors)
    }

    fn parse_nodes(&mut self) -> Vec<Node<'a>> {
        let mut nodes = Vec::new();

        while self.peek().is_some() {
            // We want to parse '\n', so keep it.
            self.consume_while(|c| c.is_whitespace() && c != '\n');

            match self.peek() {
                Some('<') => {
                    // Should be the '<' we just peeked
                    self.consume();

                    let peek = self.peek();
                    if peek.is_some_and(|c| c == '/') {
                        // Should be the '/' we just peeked.
                        self.consume();
                        self.skip_whitespace();

                        let close = self.parse_tag();

                        self.skip_whitespace();
                        self.expect_consume('>');

                        // Add to the closing tag queue
                        self.close_tag_queue.push_back(close);
                        break;
                    } else if peek.is_some_and(|c| c == '!') {
                        let start = self.slice_index;
                        let start_idx = self.index;
                        if let Some(comment) = self.try_parse_comment() {
                            nodes.push(comment);
                        } else {
                            // start - 1 is safe because we know we've consumed a < up to this
                            // point
                            let text = &self.source[(start - 1)
                                ..self.source.len() - self.source[self.slice_index..].len()];
                            nodes.push(Node::new(
                                NodeType::Text(text),
                                Location::new(start_idx - 1, self.index.saturating_sub(1)),
                            ));
                        }
                    } else {
                        let node = self.parse_element();

                        // If no closing tag was found (<br/>, for example), just push the tag
                        if self.close_tag_queue.is_empty() {
                            nodes.push(node);
                            continue;
                        }

                        let NodeType::Element(elem) = node.node_type() else {
                            unreachable!("should never parse anything other than an element")
                        };

                        // Validate that the closing tag matches the opening one
                        let tag = elem.tag();
                        let Some(close_tag) = self.close_tag_queue.pop_front() else {
                            unreachable!("checked if empty at top of block");
                        };
                        if close_tag != tag {
                            self.errors
                                .push(ParseError::InvalidClosingTag(tag.to_owned(), node.loc()));
                            nodes.push(Node::error(node.loc()));
                            continue;
                        }

                        nodes.push(node);
                    }
                }
                Some('{') => {
                    // Should be the {
                    self.consume();
                    // Accounting for the { we consumed
                    let start = self.index - 1;

                    match self.peek() {
                        Some('#') => {
                            let node = self.parse_special_block();
                            nodes.push(node);
                        }
                        Some('/') => {
                            break;
                        }
                        _ => {
                            let mustache = self.parse_mustache();
                            nodes.push(Node::new(
                                NodeType::Mustache(mustache),
                                Location::new(start, self.index - 1),
                            ))
                        }
                    }
                }
                _ => {
                    // Parse text node if not tag
                    nodes.push(self.parse_text_node());
                }
            }
        }

        nodes
    }

    fn consume_while<F>(&mut self, predicate: F) -> &'a str
    where
        F: Fn(char) -> bool,
    {
        let start = self.slice_index;
        while self.peek().is_some_and(|c| predicate(c)) {
            self.consume();
        }

        // Get the text we just consumed. Importantly, we use `self.slice_index`, as we're indexing
        // an &str, which means we could invalidate byte boundaries if we just counted by
        // `self.index`
        &self.source[start..self.source.len() - self.source[self.slice_index..].len()]
    }

    fn consume(&mut self) -> Option<char> {
        if let Some(peek) = self.peek_buffer {
            // index keeps track of how many characters we've consumed
            self.index += 1;
            self.slice_index += peek.len_utf8();
            self.peek_buffer = None;
            return Some(peek);
        }

        let c = self.input.next();
        if let Some(c) = c {
            self.index += 1;
            // Slice index should keep track of the total bytes we've consumed.
            self.slice_index += c.len_utf8();
        }
        c
    }

    fn expect_consume(&mut self, expected: char) -> bool {
        let c = self.consume();
        if !c.is_some_and(|c| c == expected) {
            self.errors.push(ParseError::ExpectedCharacter(
                expected,
                c.unwrap_or('\0'),
                Location::new(self.index, self.index),
            ));
            false
        } else {
            true
        }
    }

    fn peek(&mut self) -> Option<char> {
        if let Some(peek) = self.peek_buffer {
            return Some(peek);
        }

        let next = self.input.next()?;

        self.peek_buffer = Some(next);
        Some(next)
    }

    fn skip_whitespace(&mut self) -> &'a str {
        self.consume_while(|c| c.is_whitespace())
    }

    fn parse_text_node(&mut self) -> Node<'a> {
        self.consume_while(|c| c.is_whitespace() && c != '\n');
        let mut start = self.index;
        let mut text = self.consume_while(|c| c != '<' && c != '{');
        if text.get(0..=0).is_some_and(|c| c == "\n") {
            start = self.index.saturating_sub(1);
            text = &text[..1];
        }

        let node_type = NodeType::Text(text);
        Node::new(
            node_type,
            Location::new(start, self.index.saturating_sub(1)),
        )
    }

    fn parse_tag(&mut self) -> &'a str {
        // `is_digit(36)` checks if it's a-z, A-Z, or 0-9
        self.consume_while(|c| c.is_digit(36))
    }

    fn parse_element(&mut self) -> Node<'a> {
        let start = self.index;
        let tag = self.parse_tag();
        if tag.is_empty() {
            return Node::new(NodeType::Text("<"), Location::char(start - 1));
        }
        self.skip_whitespace();
        let attrs = self.parse_attrs();
        let children = self.parse_nodes();

        Node::new(
            NodeType::Element(Element::new(tag, attrs, children)),
            Location::new(start - 1, self.index.saturating_sub(1)),
        )
    }

    fn parse_attrs(&mut self) -> Vec<Attribute<'a>> {
        let mut attrs = vec![];

        while self.peek().is_some_and(|c| c != '>') {
            self.skip_whitespace();
            let name = self.consume_while(|c| !is_control_or_delim(c));
            match name {
                "on" => {
                    self.expect_consume(':');
                    let event = self.parse_tag();
                    self.skip_whitespace();
                    self.expect_consume('=');
                    self.skip_whitespace();
                    self.expect_consume('{');
                    let mustache = self.parse_mustache();
                    attrs.push(Attribute::EventHandler(EventHandler::new(event, mustache)));
                }
                "bind" => {
                    self.expect_consume(':');
                    let binding = self.parse_tag();
                    self.skip_whitespace();
                    attrs.push(Attribute::Binding(binding));
                }
                _ => {
                    self.skip_whitespace();
                    let value = if self.peek().is_some_and(|c| c == '=') {
                        // Consume the '='
                        self.consume();
                        self.skip_whitespace();
                        let s = self.parse_attr_value();
                        self.skip_whitespace();
                        Some(s)
                    } else {
                        None
                    };

                    attrs.push(Attribute::KeyValue(name, value));
                }
            }
        }

        // Should be the '>'
        self.consume();
        attrs
    }

    fn parse_attr_value(&mut self) -> AttributeValue<'a> {
        match self.peek() {
            Some(c) if c == '"' || c == '\'' => {
                self.consume();
                let value = self.consume_while(|str_char| str_char != c);
                self.consume();
                AttributeValue::Literal(value)
            }
            Some('{') => {
                // Should be the {
                self.consume();
                let expr = self.parse_mustache();
                AttributeValue::JavaScript(expr)
            }
            c => {
                self.errors.push(ParseError::ExpectedCharacterAny(
                    vec!['{', '"', '\''],
                    c.unwrap_or('\0'),
                    Location::char(self.index),
                ));
                AttributeValue::Literal("")
            }
        }
    }

    fn parse_mustache(&mut self) -> SyntaxNode {
        let start = self.slice_index;
        // Keep track of rbraces needed to close out mustache tag. This is needed because the inner
        // JavaScript could potentially have curly braces
        let mut rbraces_needed = 1;
        while rbraces_needed > 0 {
            let c = self.peek();
            match c {
                Some('{') => {
                    self.consume();
                    rbraces_needed += 1;
                }
                Some('}') => {
                    rbraces_needed -= 1;
                    if rbraces_needed != 0 {
                        self.consume();
                    }
                }
                None | Some(_) => {
                    self.consume();
                }
            }
        }

        // This is not extracted into a function, because it is very much tied to the current state
        // of the parser. A misuse of this could be bad and hard to debug
        let expr = &self.source[start..self.source.len() - self.source[self.slice_index..].len()];

        // Should be the last rbrace needed
        self.expect_consume('}');

        // Turn into rslint JavaScript syntax tree
        rslint_parser::parse_text(expr, 0).syntax()
    }

    fn try_parse_comment(&mut self) -> Option<Node<'a>> {
        // Start includes the < already consumed
        let start = self.index - 1;
        let (_, found) = self.try_consume("!--");
        if !found {
            // Returns None so calling function knows that there was no comment. A text node should
            // be created instead
            return None;
        }

        let start_slice = self.slice_index;
        let mut found = false;
        while !found && self.peek().is_some() {
            self.consume_while(|c| c != '-');
            let (_, successful) = self.try_consume("-->");
            found = successful;
        }

        let text =
            &self.source[start_slice..self.source.len() - self.source[self.slice_index..].len()];
        if !text.ends_with("-->") {
            self.errors
                .push(ParseError::ExpectedEndOfComment(Location::new(
                    start,
                    self.index - 1,
                )));
            return None;
        }
        Some(Node::new(
            // This slice is safe because we checked the above
            NodeType::Comment(&text[..text.len() - 3]),
            Location::new(start, self.index - 1),
        ))
    }

    fn try_consume(&mut self, text: &str) -> (&'a str, bool) {
        let start = self.slice_index;
        let pieces_needed = text.chars();
        for needed in pieces_needed {
            if !self.consume().is_some_and(|c| c == needed) {
                let t =
                    &self.source[start..self.source.len() - self.source[self.slice_index..].len()];
                return (t, false);
            }
        }

        (
            &self.source[start..self.source.len() - self.source[self.slice_index..].len()],
            true,
        )
    }

    fn parse_special_block(&mut self) -> Node<'a> {
        let start = self.index - 1;
        self.expect_consume('#');
        let block_type = self.consume_while(|c| !c.is_whitespace() && c != '}');

        match block_type {
            "for" => {
                let mut error = false;

                self.skip_whitespace();
                let mut ident =
                    self.consume_while(|c| !c.is_whitespace() && c != ',' && c.is_digit(36));
                self.skip_whitespace();
                let mut index_ident = None;
                if self.peek().is_some_and(|c| c == ',') {
                    // Should be the ','
                    self.consume();
                    self.skip_whitespace();
                    // First one found should be index ident
                    index_ident = Some(ident);
                    ident = self.consume_while(|c| !c.is_whitespace() && c.is_digit(36));
                }
                self.skip_whitespace();
                {
                    let start = self.index;
                    let (got, found) = self.try_consume("in");
                    if !found {
                        let mut full_got = got.to_owned();
                        full_got.push_str(self.consume_while(|c| !c.is_whitespace() && c != '}'));
                        self.errors.push(ParseError::ExpectedString(
                            "in".to_owned(),
                            full_got,
                            Location::new(start, self.index),
                        ));
                        error = true;
                    }
                }
                self.skip_whitespace();
                let expr = self.parse_mustache();
                let inner = self.parse_nodes();
                self.expect_consume('/');
                {
                    let start = self.index;
                    let (got, found) = self.try_consume(block_type);
                    if !found {
                        let mut full_got = got.to_owned();
                        full_got.push_str(self.consume_while(|c| c != '}'));
                        self.errors.push(ParseError::InvalidClosingBlock(
                            "for".to_owned(),
                            full_got,
                            Location::new(start, self.index),
                        ));
                        error = true;
                    }
                }
                self.expect_consume('}');
                if error {
                    Node::error(Location::new(start, self.index))
                } else {
                    Node::new(
                        NodeType::SpecialBlock(SpecialBlock::For(ForBlock::new(
                            ident,
                            index_ident,
                            expr,
                            inner,
                        ))),
                        Location::new(start, self.index),
                    )
                }
            }
            "if" => {
                let mut error = false;
                self.skip_whitespace();
                let expr = self.parse_mustache();
                let inner = self.parse_nodes();
                self.expect_consume('/');
                {
                    let start = self.index;
                    let (got, found) = self.try_consume(block_type);
                    if !found {
                        let mut full_got = got.to_owned();
                        full_got.push_str(self.consume_while(|c| c != '}'));
                        self.errors.push(ParseError::InvalidClosingBlock(
                            "if".to_owned(),
                            full_got,
                            Location::new(start, self.index),
                        ));
                        error = true;
                    }
                }
                self.expect_consume('}');
                if error {
                    Node::error(Location::new(start, self.index))
                } else {
                    Node::new(
                        NodeType::SpecialBlock(SpecialBlock::If(IfBlock::new(expr, inner, None))),
                        Location::new(start, self.index),
                    )
                }
            }
            _ => {
                self.consume_while(|c| c != '}');
                self.consume();
                self.consume_while(|c| c != '}');
                self.consume();
                self.errors.push(ParseError::UnrecognizedSpecialBlock(
                    block_type.to_owned(),
                    Location::new(start, self.index),
                ));
                Node::error(Location::new(start, self.index))
            }
        }
    }
}

fn is_control_or_delim(ch: char) -> bool {
    match ch {
        '\u{007F}' => true,
        c if c >= '\u{0000}' && c <= '\u{001F}' => true,
        c if c >= '\u{0080}' && c <= '\u{009F}' => true,
        ' ' | '"' | '\'' | '>' | '/' | '=' | ':' => true,
        _ => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! insta_test_all {
        ($($input:expr),+) => {
            $(
                let parser = Parser::new($input);
                let (tree, errs) = parser.parse();
                assert_eq!(Vec::<ParseError>::new(), errs);
                ::insta::assert_debug_snapshot!(tree);
             )+
        };
    }

    macro_rules! insta_test_all_err {
        ($($input_and_errs:expr),+) => {
            $(
                let parser = Parser::new($input_and_errs.0);
                let (tree, errs) = parser.parse();
                assert_eq!($input_and_errs.1, errs);
                ::insta::assert_debug_snapshot!(tree);
             )+
        };
    }

    #[test]
    fn can_parse_text_node() {
        insta_test_all!(
            "hello",
            "   hello",
            "hello hello",
            "你",
            "你好",
            "<",
            ">",
            "<>hello"
        );
    }

    #[test]
    fn can_parse_basic_elements() {
        insta_test_all!(
            "<p>你好</p>",
            "<tagname><p>inner</p></tagname>",
            "<p>
                <p>hello</p>
            </p>
            "
        );
    }

    #[test]
    fn parses_basic_attrs() {
        insta_test_all!(
            "<p attr1='hello'></p>",
            "<p attr=\"attribute\"></p>",
            "<p ></p>",
            "<p attr=\"你好\"></p>",
            "<p attr=\"one\"     attr2=\"two\">text</p>",
            "<p attr></p>"
        );
    }

    #[test]
    fn reports_unmatched_tag_errors() {
        insta_test_all_err!((
            "<p></invalid>",
            vec![ParseError::InvalidClosingTag(
                "p".to_owned(),
                Location::new(0, 12)
            )]
        ));
    }

    #[test]
    fn continues_parsing_if_unmatched_tag_found() {
        insta_test_all_err!(
            (
                "<p></invalid>
            <span>hello</span>",
                vec![ParseError::InvalidClosingTag(
                    "p".to_owned(),
                    Location::new(0, 12)
                )],
            ),
            (
                "
                <span>hello
                <p></p>
                <span></span>
                ",
                Vec::<ParseError>::new()
            )
        );
    }

    #[test]
    fn reports_invalid_closing_tag_angle_bracket() {
        insta_test_all_err!((
            "<p></p!",
            vec![ParseError::ExpectedCharacter('>', '!', Location::new(7, 7))]
        ));
    }

    #[test]
    fn can_parse_mustache_in_event_handle_attr() {
        insta_test_all!(
            "<p on:click={hello}></p>",
            "<p on:click={ () => { console.log(\"hi\") } }></p>"
        );
    }

    #[test]
    fn can_parse_bindings() {
        insta_test_all!("<p bind:value></p>");
    }

    #[test]
    fn can_parse_attrs_with_mustache_values() {
        insta_test_all!(
            "<p attr={value}></p>",
            "<p attr={ () => { console.log(value) } }></p>"
        );
    }

    #[test]
    fn can_parse_comments() {
        insta_test_all!("<!-- hello -->", "<!--hello--><p></p>", "<!-");
    }

    #[test]
    fn gives_error_when_end_of_comment_not_found() {
        insta_test_all_err!(
            (
                "<!-- hello",
                vec![ParseError::ExpectedEndOfComment(Location::new(0, 9))]
            ),
            (
                "<!-- hello<p>hello</p>",
                vec![ParseError::ExpectedEndOfComment(Location::new(0, 21))]
            )
        );
    }

    #[test]
    fn can_parse_mustache_tags() {
        insta_test_all!(
            "   {hello}",
            "{ () => { console.log(\"HI\") } }",
            "hello, {hello}"
        );
    }

    #[test]
    fn can_parse_if_block() {
        insta_test_all!("{#if x == 3}<p>hello</p>{/if}");
    }

    #[test]
    fn can_parse_for_block() {
        insta_test_all!(
            "{#for i in [1, 2, 3]}<p>{i}</p>{/for}",
            "{#for i, elem in [1, 2, 3]}<p>Index: {i}. Elem: {elem}</p>{/for}",
            "{#for i in [1, 2, 3]}<p>{i}</p>{/for}
            <p>Text</p>",
            "{#for i in [1, 2, 3]}
                {#if i == 1}
                    <span>{i}</span>
                {/if}
            {/for}"
        );
    }

    #[test]
    fn errors_on_invalid_closing_special_block() {
        insta_test_all_err!(
            (
                "{#for i in [1, 2, 3]}{i}{/if}",
                vec![ParseError::InvalidClosingBlock(
                    "for".to_owned(),
                    "if".to_owned(),
                    Location::new(26, 28)
                )]
            ),
            (
                "{#if i == 3}{i}{/for}",
                vec![ParseError::InvalidClosingBlock(
                    "if".to_owned(),
                    "for".to_owned(),
                    Location::new(17, 20)
                )]
            )
        );
    }

    #[test]
    fn error_when_no_in_token_found_in_for_block() {
        insta_test_all_err!((
            "{#for i bin [1, 2, 3]}{i}{/for}",
            vec![ParseError::ExpectedString(
                "in".to_owned(),
                "bin".to_owned(),
                Location::new(8, 11)
            )]
        ));
    }

    #[test]
    fn error_on_unrecognized_special_block() {
        insta_test_all_err!((
            "{#what}stuff{/what}",
            vec![ParseError::UnrecognizedSpecialBlock(
                "what".to_owned(),
                Location::new(0, 19)
            )]
        ));
    }
}
