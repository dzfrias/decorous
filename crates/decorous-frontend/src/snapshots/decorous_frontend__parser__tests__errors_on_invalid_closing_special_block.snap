---
source: crates/decorous-frontend/src/parser.rs
expression: tree
---
[
    Node {
        loc: Location {
            start: 0,
            end: 29,
        },
        node_type: Error,
    },
]
